fendermint/vm/topdown/src/sync.rs

```diff
@@ -0,0 +1,205 @@
+// Copyright 2022-2023 Protocol Labs
+// SPDX-License-Identifier: Apache-2.0, MIT
+//! A constant running process that fetch or listener to parent state
+
+use crate::error::Error;
+use crate::{BlockHash, BlockHeight, Config, InMemoryFinalityProvider, ParentFinalityProvider};
+use anyhow::Context;
+use async_stm::atomically_or_err;
+use fvm_shared::clock::ChainEpoch;
+use ipc_agent_sdk::apis::IpcAgentClient;
+use ipc_agent_sdk::jsonrpc::JsonRpcClientImpl;
+use ipc_sdk::cross::CrossMsg;
+use ipc_sdk::subnet_id::SubnetID;
+use ipc_sdk::ValidatorSet;
+use std::sync::Arc;
+use std::time::Duration;
+
+/// Constantly syncing with parent through polling
+pub struct PollingParentSyncer {
+    config: Config,
+    parent_view_provider: Arc<InMemoryFinalityProvider>,
+    agent: Arc<IPCAgentProxy>,
+}
+
+impl PollingParentSyncer {
+    pub fn new(
+        config: Config,
+        parent_view_provider: Arc<InMemoryFinalityProvider>,
+        agent: Arc<IPCAgentProxy>,
+    ) -> Self {
+        Self {
+            config,
+            parent_view_provider,
+            agent,
+        }
+    }
+}
+
+impl PollingParentSyncer {
+    /// Start the parent finality listener in the background
+    pub fn start(self) -> anyhow::Result<()> {
+        let config = self.config;
+        let provider = self.parent_view_provider;
+        let agent = self.agent;
+
+        tokio::spawn(async move {
+            loop {
+                if let Err(e) = sync_with_parent(&config, &agent, &provider).await {
+                    tracing::info!("sync with parent encountered error: {e}");
+                }
+            }
+        });
+
+        Ok(())
+    }
+}
+
+/// Extract the actual error from Box
+macro_rules! downcast_err {
+    ($r:expr) => {
+        match $r {
+            Ok(v) => Ok(v),
+            Err(e) => match e.downcast_ref::<Error>() {
+                None => unreachable!(),
+                Some(e) => Err(e.clone()),
+            },
+        }
+    };
+}
+
+async fn sync_with_parent<T: ParentFinalityProvider + Send + Sync + 'static>(
+    config: &Config,
+    agent_proxy: &Arc<IPCAgentProxy>,
+    provider: &Arc<T>,
+) -> anyhow::Result<()> {
+    let mut interval = tokio::time::interval(Duration::from_secs(config.polling_interval_secs));
+
+    loop {
+        interval.tick().await;
+
+        let (starting_height, latest_height) =
+            get_query_block_heights(agent_proxy, provider).await?;
+        if latest_height < config.chain_head_delay {
+            tracing::debug!("latest height not more than the chain head delay");
+            continue;
+        }
+
+        // we are going backwards in terms of block height, the latest block height is lower
```

Comment:
While this makes sense from the syncing/caching perspective, what may be a bit more complicated is how to handle all the on-chain side-effects that may have happened before the reorg. Finality commitments and cross-net messages may need to be reverted. This may be hard to handle and not that obvious :(